packages/upload-media/src/store/utils/vips.ts

```diff
@@ -0,0 +1,113 @@
+/**
+ * External dependencies
+ */
+import { createWorkerFactory } from '@shopify/web-worker';
```

Comment:
I'm a bit skeptical about this. I have no big expertise with web workers but this is yet another thing that will force npm package users to have a special bundler config. It might even force them to use webpack. 

The question is the same, how do regular npm packages ship and use web workers, is this even a thing or are web workers limited to application code.

Reply:
Yeah the web worker part will be a challenge to figure out. I might remove it from this PR for now.

This package here is a great fit in theiry with good DX, but it does indeed require some wiring. Plus, as of very recently it‘s now in maintenance mode, see https://github.com/swissspidy/media-experiments/issues/713 🙃

Loading this in a web worker is a requirement from a performance perspective, and it needs to be done in a way that works in a WordPress context (i.e. no control over the server stack) and without affecting the main bundle.

Do you have anyone on your team who could help evaluate alternatives? :-)

Reply:
We do have some good people on the team but the problem is the time and the availability. Also the challenge we have to ship a block editor for WordPress but also for the web as a framework (npm package) make this kind of things very challenging.

cc @gziolo @sirreal

Reply:
I'll have a closer look next week. Thank you for ping, @youknowriad.

Reply:
With this open question regarding the web worker — would you prefer completely removing this part of the code for now (even though it's not a published package) or is it OK to keep for now so we can further iterate on it (since it's behind an experiment anyway)?

Reply:
Personally this is my main worry for this PR. I wonder if we should leave all the wasm and web worker imports out of the initial work. I now all of this is experimental but there are non experimental npm packages that will depend on these packages.

Reply:
Only `upload-media` would depend on them, which is experimental.

As long as this doesn't affect npm publishing and, that should be fine? Ideally someone installing `@wordpress/block-editor` in their project shouldn't end up installing them. If we can achieve that, that would be great.

Reply:
since `@wordpress/block-editor` depends on `upload-media`, it's the same issue for people already using the block-editor in their apps no? 

Agreed that the ideal is that there's some way to opt-in without breaking the default. 
The other thing to think about is when these things ship in Core, they become public APIs that Core won't be able to break. (so handling them like we handle jQuery and React upgrades) and this is something we need to minimize a lot. I don't see us doing this for small npm packages for instance.

Reply:
`block-editor` does not actually depend on `upload-media` yet. This PR does not add this connection (not even #66380 does that). So right now there shouldn't be anything to worry about and we can figure this out later on. Is there a problem with that?

> The other thing to think about is when these things ship in Core, they become public APIs that Core won't be able to break

That all makes sense to me, but can't this be figured out in future PRs? I was asked to split this up into multiple PRs for a reason. :-)

Reply:
> That all makes sense to me, but can't this be figured out in future PRs? I was asked to split this up into multiple PRs for a reason. :-)

The main issue is the way the PRs are split I think. historically, we try to split it functionally. Start with a small feature and expand the feature. This PR is trying to ship a full featured npm package that is not used anywhere and then discover the flaws later when we integrate it. It's just not the way we've approached things so far and makes future PRs very hard to review/test properly because the scale of the impact will be unknown really.

If others feel confident shipping this PR, I'm not blocking it personally but I would say in my mind, it would have been better to introduce this as a working feature with the smallest `upload-media` possible if that makes sense.

Happy to defer to others though. cc @WordPress/gutenberg-core

Reply:
I get where you're coming from, and that's the usual way when you build something from scratch.

However, here I am porting over [existing code from a battle-tested plugin](https://github.com/swissspidy/media-experiments), that I have asked GB contributors for over a year to use, test, and provide feedback for. I have outlined the proposed "modus operandi" both in personal discussions as well as tickets such as #61447, precisely because it's an existing codebase and taking it apart to reassemble it step by step would be very inefficient. Receiving this feedback so late in the process is a bit frustrating, I hope you understand that.

That said, if you think stripping this PR back to the minimum and reassembling everything later is the best course of action, I can certainly do that. But then I kindly ask you once again to also look at the [existing code base](https://github.com/swissspidy/media-experiments) so you see the full picture and better understand how everything is meant to work together.

Reply:
I don't have much context here, I'm new to this. But I have to agree with functional changes, it seems a bit odd to add a package without any uses and figuring it out later. We don't normally take code from another plugin and then add it bit by bit without re-evaluating it all. That's the same for WordPress core I believe. Why don't we take the simplest feature and add it here? It would be much easier to test and understand how it works.

Reply:
I totally understand where you're coming from as well but for me personally at least, I think the plugin should have been built in Gutenberg from the start.

I don't have the capacity to review such a big code base at once. I'm happy if others that help you in the project do though. I wish I can help more, I can give you an approval without checking but I feel like I'll be doing you both of us a disservice. 

This reminds of the "inline comments" work as well. MultiDots have a full featured plugin as well that they wanted to contribute in Core and the approach there was also to scale down the plugin and introduce it bit by bit.

Reply:
That plugin came to be through rapid prototyping and numerous refactorings and epoeriments. It also contains some parts not intended for core. That would have been **impossible** to build in Gutenberg from the start. But that‘s not the point right now.

I will strip down the PR and then request another review.

> later. We don't normally take code from another plugin and then add it bit by bit without re-evaluating it all. That's the same for WordPress core I believe.

There are feature plugins, FWIW

> It would be much easier to test and understand how it works.

I recommend using the existing plugin in the meantime to do that. At least _using_ the plugin in Playground shouldn‘t be too much to ask.

Reply:
> There are feature plugins, FWIW

Yes, I know, I've made some in the past. :) As far as I know none of them are just absorbed into core like this though?

> I recommend using the existing plugin in the meantime to do that. At least using the plugin in Playground shouldn‘t be too much to ask.

I mean testing this branch. And seeing _these_ changes in action through a use case.